Return false if validation fails in `commonAssignmentCheck`

aosen-xiong opened the pull request.

aosen-xiong committed:
Return false if validation fail

wmdietl commented:
Is any code actually using the returned value of `commonAssignmentCheck` and `checkArrayInitialization`? Can you look whether any subclasses actually use the result?
If there is a use, can you write a test case that fails before this change?

aosen-xiong committed:
Merge branch 'master' into fix-return

aosen-xiong committed:
Merge branch 'master' into fix-return

aosen-xiong committed:
Merge branch 'master' into fix-return

aosen-xiong committed:
Add regression test for commonAssignmentCheck contract

Adds a small dedicated test checker that triggers the validateType failure path of BaseTypeVisitor.commonAssignmentCheck(Tree, ExpressionTree, ...). The visitor's commonAssignmentCheck override emits a sentinel warning only when super returns true, which surfaces the contract violation: before the fix, super returned true after a failed validateType, causing the warning to be issued; after the fix, super correctly returns false.

aosen-xiong committed:
TEMP: revert fix to demonstrate test failure in CI

aosen-xiong committed:
Restore fix and add SKIP-REQUIRE-JAVADOC

Restores the return false fix in commonAssignmentCheck after the temporary revert that demonstrated the test failure. Also adds SKIP-REQUIRE-JAVADOC to skip the requireJavadoc check while this PR is in progress.

wmdietl reviewed (commented):
Thanks for the extensive test. It seems like a very specialized test. I'm not sure we need to keep it around. Do you think we'll have any use for it in the future?
Is there no existing checker where we could cause a type.invalid error?

aosen-xiong committed:
Merge branch 'master' into fix-return

aosen-xiong committed:
Reuse h1h2 checker for assignment validation test

aosen-xiong committed:
Fix h1h2 diagnostic comment placement

aosen-xiong committed:
Simplify common assignment regression sentinel

aosen-xiong committed:
Document H1H2 type validator

wmdietl commented on framework/tests/h1h2checker/TypeRefinement.java:
```diff
@@ -4,15 +4,14 @@
 public class TypeRefinement {
     // :: warning: (cast.unsafe.constructor.invocation)
     @H1Top Object o = new @H1S1 Object();
-    // :: error: (h1h2checker.h1invalid.forbidden) :: warning: (cast.unsafe.constructor.invocation)
```
Does it make sense that the `cast.unsafe.constructor.invocation` warning is gone?

wmdietl reviewed (commented):
Nice way to test it! One question about a warning we no longer get.

aosen-xiong committed:
Preserve h1 invalid constructor warning

aosen-xiong committed:
Merge branch 'master' into fix-return

aosen-xiong committed:
Document H1H2 visitor

aosen-xiong committed:
Merge branch 'master' into fix-return

aosen-xiong committed:
Return false for invalid assignment value types

aosen-xiong committed:
Merge branch 'master' into fix-return
